server/permalinks.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"context"
+	"encoding/hex"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/v25/github"
+)
+
+const maxPermalinkReplacements = 10
+const permalinkReqTimeout = 2 * time.Second
+const maxPreviewLines = 10
+
+// replacement holds necessary info to replace github permalinks
+// in messages with a code preview block
+type replacement struct {
+	index      int               // index of the permalink in the string
+	word       string            // the permalink
+	captureMap map[string]string // named regex capture group of the link
+}
+
+// getReplacements returns the permalink replacements that needs to be performed
+// on a message.
+func (p *Plugin) getReplacements(msg string) []replacement {
+	// find the permalinks from the msg using a regex
+	matches := p.githubPermalinkRegex.FindAllStringSubmatch(msg, -1)
+	indices := p.githubPermalinkRegex.FindAllStringIndex(msg, -1)
+	replacements := make([]replacement, 0, maxPermalinkReplacements)
+	for i, m := range matches {
+		// have a limit on the no. of replacements to do
+		if i > maxPermalinkReplacements {
+			break
+		}
+		word := m[0]
+		index := indices[i][0]
+		r := replacement{
+			index: index,
+			word:  word,
+		}
+		// ignore if the word is inside a link
+		if isInsideLink(msg, index) {
+			continue
+		}
+		// populate the captureMap with the extracted groups of the regex
+		r.captureMap = make(map[string]string)
+		for j, name := range p.githubPermalinkRegex.SubexpNames() {
+			if j == 0 {
+				continue
+			}
+			r.captureMap[name] = m[j]
+		}
+		replacements = append(replacements, r)
+	}
+	return replacements
+}
+
+// makeReplacements perform the given replacements on the msg and returns
+// the new msg.
+func (p *Plugin) makeReplacements(msg string, replacements []replacement) string {
+	for _, r := range replacements {
+		// quick bailout if the commit hash is not proper
+		if _, err := hex.DecodeString(r.captureMap["commit"]); err != nil {
+			p.API.LogError("bad git commit hash in permalink", "error", err.Error(), "hash", r.captureMap["commit"])
+			continue
+		}
+
+		// get the file contents
+		opts := github.RepositoryContentGetOptions{
+			Ref: r.captureMap["commit"],
+		}
+		// TODO: make all of these requests concurrently.
+		reqctx, cancel := context.WithTimeout(context.Background(), permalinkReqTimeout)
+		fileContent, _, _, err := p.githubClient.Repositories.GetContents(reqctx,
```

Comment:
Also it looks like `GetContents` has a max 1mb limit, we should probably make sure we return something valid in cases where we can't find the start/end from the file.

https://github.com/google/go-github/blob/0830f2846d799e0c1b07b643f79c4a2af7e54295/github/repos_contents.go#L114

Reply:
Yes, I had mentioned this in my design sketch. I think it is good to have this limit to prevent unnecessarily downloading huge files just to show a 2 line preview. In this case, it will just not expand the link, that's all.